lib/engine-jit/src/artifact.rs

```diff
@@ -325,15 +331,51 @@ impl Artifact for JITArtifact {
         &self.func_data_registry
     }
     fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
-        // let mut s = flexbuffers::FlexbufferSerializer::new();
-        // self.serializable.serialize(&mut s).map_err(|e| SerializeError::Generic(format!("{:?}", e)));
-        // Ok(s.take_buffer())
-        let bytes = bincode::serialize(&self.serializable)
-            .map_err(|e| SerializeError::Generic(format!("{:?}", e)))?;
-
         // Prepend the header.
         let mut serialized = Self::MAGIC_HEADER.to_vec();
-        serialized.extend(bytes);
+
+        serialized.resize(SERIALIZED_METADATA_CONTENT_OFFSET, 0);
+        let mut writable_leb = &mut serialized[SERIALIZED_METADATA_LENGTH_OFFSET..];
+        let serialized_data = self.serializable.serialize()?;
+        let length = serialized_data.len();
+        leb128::write::unsigned(&mut writable_leb, length as u64).expect("Should write number");
+
+        let offset = pad_and_extend::<SerializableModule>(&mut serialized, &serialized_data);
+        assert_eq!(offset, SERIALIZED_METADATA_CONTENT_OFFSET);
+
         Ok(serialized)
     }
 }
+
+/// It pads the data with the desired alignment
+pub fn pad_and_extend<T>(prev_data: &mut Vec<u8>, data: &[u8]) -> usize {
+    let align = std::mem::align_of::<T>();
+
+    let mut offset = prev_data.len();
+    if offset & (align - 1) != 0 {
+        offset += align - (offset & (align - 1));
+        prev_data.resize(offset, 0);
+    }
+    prev_data.extend(data);
+    offset
+}
+
+#[cfg(test)]
+mod tests {
+    use super::pad_and_extend;
+
+    #[test]
+    fn test_pad_and_extend() {
+        let mut data: Vec<u8> = vec![];
+        let offset = pad_and_extend::<i64>(&mut data, &[1, 0, 0, 0, 0, 0, 0, 0]);
+        assert_eq!(offset, 0);
+        let offset = pad_and_extend::<i32>(&mut data, &[2, 0, 0, 0]);
+        assert_eq!(offset, 8);
+        let offset = pad_and_extend::<i64>(&mut data, &[3, 0, 0, 0, 0, 0, 0, 0]);
+        assert_eq!(offset, 16);
```

Comment:
Similarly, I'd expect this to be 8, not 16

Reply:
Oh nevermind, I misunderstood, I got confused between the slice after `data` and `data` itself. This looks correct 👍